torch/lib/c10d/reducer.cpp

```diff
@@ -1159,6 +1177,7 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
           local_used_work_->wait();
           // D2H from local_used_maps_dev_ to local_used_maps_
           for (size_t i = 0; i < local_used_maps_.size(); i++) {
+            // Blocking copy, if local_used_maps_dev_ is cuda
```

Comment:
Just for confirming whether I'm understanding this issue correctly. 

So if at some iteration, all params are used in some ranks and blocking copy will not be called here, host and device will not be synced in this case and thus result in race condition as you mentioned for those ranks.  

but if there is at least one param is not used in the ranks, host and device will be synced here and thus results should be fine.

Reply:
> if there is at least one param is not used in the ranks, host and device will be synced here and thus results should be fine.

Danger is determined within each rank, as you said in your second sentence ("thus result in race condition as you mentioned for those ranks"). Within a given iteration, any rank that did not use all params (and calls the `Blocking copy`) is not in danger.

Any other rank that _did_ use all parameters (even in the same iteration) is still in danger, because it can skip the blocking copy and proceed all the way to `local_used.fill_(0);` without blocking.

Reply:
right, your explanation is more clear

Reply:
Haha your understanding was correct, i just wanted to be super meticulous about phrasing because this is a tricky issue.